ass/document.py

```diff
@@ -244,165 +246,12 @@ def from_ass(cls, s):
-                    # Style: ...
-                    if type_name.lower() != Style.TYPE.lower():
-                        raise ValueError("expected style line in styles")
-
-                    doc.styles.append(Style.parse(line, field_order))
-
-            elif section.lower() == doc.EVENTS_HEADER.lower():
-                # [Events]
-                if ':' not in line:
-                    continue
-
-                type_name, _, line = line.partition(":")
-                line = line.lstrip()
-
-                if field_order is None:
-                    # Format: ...
-                    if type_name.lower() != Document.FORMAT_TYPE.lower():
-                        raise ValueError("expected format line in events")
-
-                    field_order = [field.strip() for field in line.split(",")]
-                    doc.events_field_order = field_order
-                else:
-                    # Dialogue: ...
-                    # Comment: ...
-                    # etc.
-                    doc.events.append(({
-                        "Dialogue": Dialogue,  # noqa: E241
-                        "Comment":  Comment,   # noqa: E241
-                        "Picture":  Picture,   # noqa: E241
-                        "Sound":    Sound,     # noqa: E241
-                        "Movie":    Movie,     # noqa: E241
-                        "Command":  Command    # noqa: E241
-                    })[type_name].parse(line, field_order))
-
-            else:
-                # unknown sections
-                continue
-
-        return doc
-
-    def dump_file(self, f):
-        """ Dump this ASS document to a file object.
-        """
-        if getattr(f, 'encoding', 'utf_8_sig') != 'utf_8_sig':
-            import warnings
-            warnings.warn("It is recommended to write UTF-8 with BOM"
-                          " using the 'utf_8_sig' encoding")
-
-        f.write(Document.SCRIPT_INFO_HEADER + "\n")
-        for k in itertools.chain(
-            (field for field in self.DEFAULT_FIELD_ORDER if field in self.fields),
-            (field for field in self.fields if field not in self._field_mappings),
-        ):
-            f.write(k + ": " + _Field.dump(self.fields[k]) + "\n")
-        f.write("\n")
-
-        f.write(Document.STYLE_ASS_HEADER + "\n")
-        f.write(Document.FORMAT_TYPE + ": " + ", ".join(self.styles_field_order) + "\n")
-        for style in self.styles:
-            f.write(style.dump_with_type(self.styles_field_order) + "\n")
-        f.write("\n")
-
-        f.write(Document.EVENTS_HEADER + "\n")
-        f.write(Document.FORMAT_TYPE + ": " + ", ".join(self.events_field_order) + "\n")
-        for event in self.events:
-            f.write(event.dump_with_type(self.events_field_order) + "\n")
-        f.write("\n")
-
-
 @add_metaclass(_WithFieldMeta)
 class _Line(object):
-    def __init__(self, *args, **kwargs):
+    # to be overridden in subclasses or through the type_name argument
+    TYPE = None
```

Comment:
If this can be overridden in the instance through `__init__` then it mustn't be in all-caps because that's reserved for constants.

However, to change that we'd have to do a more large-scale refactor. I am willing to do that eventually, because the current architecture doesn't convince me as "good", but I'd need more time to think about this than a simple review. For now, this should be fine, since it maintains backwards compatibility.

Reply:
Yes, the reason was mostly backwards compatibility here.